### PR TITLE
Snapshot

### PR DESCRIPTION
**Pull request by themanforfree**
described in #176 and based on #185

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
implement snapshot
* what changes does this pull request make?
add snapshot api, implement for MemoryEngine and RocksRngine
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

**Comment by rogercloud** [comment]
Many interfaces are almost the same as #185. Please deduplicate.

**Review by Phoenix500526 — DISMISSED** [review]

File: engine/src/engine_api.rs [hunk 1]
```diff
@@ -87,7 +101,27 @@ pub trait StorageEngine: Send + Sync + 'static + std::fmt::Debug {
     /// flag is true, writes will be slower.
     ///
     /// # Errors
-    /// Return `EngineError::TableNotFound` if the given table does not exist
-    /// Return `EngineError` if met some errors
+    /// Return `TableNotFound` if the given table does not exist
+    /// Return `IoError` if met some io errors
```
Comment (markcty) [inline engine/src/engine_api.rs:105]: It's actually Engine Error?

**Comment by markcty** [comment]
Plz add `Send+Sync+Seek` for `SnapshotApi`.

**Review by rogercloud — COMMENTED** [review]
Currently the I/O operations are all syncronized operation, can we use `async` for better performance (sometime not so obviously)?

Yes, we can delay this part, but introducing async changes the function signature and the calling  function. I recommend introduce async as earlier as possible.

File: engine/src/rocksdb_engine.rs [hunk 2]
```diff
@@ -60,7 +67,357 @@ impl RocksEngine {
...
+            }
+        }
+        Ok(0)
+    }
+}
+
+impl Write for RocksSnapshot {
+    #[inline]
+    #[allow(clippy::indexing_slicing)] // safe indexing
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

File: engine/src/rocksdb_engine.rs [hunk 3]
```diff
@@ -135,6 +492,62 @@ impl StorageEngine for RocksEngine {
...
+                return Err(EngineError::TableNotFound((*cf_name).to_owned()));
+            };
+            let iter = snap.iterator_cf(&cf_handle, IteratorMode::Start);
+            for r in iter {
+                let (key, value) = r?;
+                if let Some(ref mut sst_writer) = sst_writer_option {
+                    sst_writer.put(key, value)?;
+                } else {
+                    let mut sst_writer = SstFileWriter::create(&opts);
+                    let file_name = format!("{cf_name}.sst");
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs:582]: cf_name is a member of tabels, and tables is const

File: engine/src/rocksdb_engine.rs [hunk 4]
```diff
@@ -60,7 +67,357 @@ impl RocksEngine {
...
+        let dir = dir.into();
+        let files = fs::read_dir(&dir)?;
+        let mut s = Self::new(dir)?;
+        for file in files {
+            let entry = file?;
+            let filename = entry.file_name().into_string().map_err(|_e| {
+                EngineError::InvalidArgument("cannot convert filename to string".to_owned())
+            })?;
+            let size = entry.metadata()?.len();
+            s.snap_files.push(SnapFile {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs:175]: we just need latest one, and here is not different versions of snapshots, here is different files of one snapshot

File: engine/src/engine_api.rs [hunk 5]
```diff
@@ -111,4 +125,24 @@ pub trait StorageEngine: Send + Sync + 'static + std::fmt::Debug {
     /// Return `EngineError::TableNotFound` if the given table does not exist
     /// Return `EngineError` if met some errors
     fn write_batch(&self, wr_ops: Vec<WriteOperation>, sync: bool) -> Result<(), EngineError>;
+
+    /// Get a snapshot of the current state of the database
+    ///
+    /// # Errors
+    /// Return `EngineError` if met some errors when creating the snapshot
+    fn snapshot(
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/engine_api.rs]: renamed

File: engine/src/rocksdb_engine.rs [hunk 6]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+    /// path of current file
+    fn current_file_path(&self) -> PathBuf {
+        let Some(current_filename) = self.snap_files.get(self.file_index).map(|sf| &sf.filename) else {
+            unreachable!("this method must be called when self.file_index < self.snap_files.len()")
+        };
+        self.dir.join(current_filename)
+    }
+
+    /// tmp path of current file
+    fn current_file_tmp_path(&self) -> PathBuf {
```
Comment (Phoenix500526) [inline engine/src/rocksdb_engine.rs:228]: The `current_file_tmp_path` is very similar to the `current_file_path`. I think we'd better merge the together.
```suggestion
    fn current_file_path(&self, tmp: bool) -> PathBuf {
        let Some(current_filename) = self.snap_files.get(self.file_index).map(|sf| &sf.filename) else {
            unreachable!("this method must be called when self.file_index < self.snap_files.len()")
        };
        if tmp {
            self.dir.join(current_filename)
        } else {
            self.dir.join(format!("{current_filename}.tmp"))
        }
    }
```

**Review by Phoenix500526 — COMMENTED** [review]
In implementing RocksSanpshot's `AsyncRead` and `AsyncWrite` traits, some code is used to process the index. I think we can use the [bytes](https://docs.rs/bytes/1.4.0/bytes/) crate to simplify these operations.

File: engine/src/memory_engine.rs [hunk 7]
```diff
@@ -206,4 +308,18 @@ mod test {
...
     }
+
+    #[test]
+    fn snapshot_should_work() {
+        let engine = MemoryEngine::new(&TESTTABLES).unwrap();
+        let put = WriteOperation::new_put("kv", "key", "value");
+        assert!(engine.write_batch(vec![put], false).is_ok());
+
+        let snapshot = engine.get_snapshot("", &TESTTABLES).unwrap();
+        let engine_2 = MemoryEngine::new(&TESTTABLES).unwrap();
```
Comment (Phoenix500526) [inline engine/src/memory_engine.rs:319]: On test coverage grounds, I think we can put another key-value pair into the original engine, and make sure that the new put operation after taking a snapshot won't show up in engine_2.

File: engine/src/rocksdb_engine.rs [hunk 8]
```diff
@@ -253,4 +729,77 @@ mod test {
...
+        assert!(engine.write_batch(vec![put], false).is_ok());
+
+        let mut snapshot = engine.get_snapshot(&snapshot_dir, &TESTTABLES).unwrap();
+        let mut buf = Vec::new();
+        snapshot.read_to_end(&mut buf).await.unwrap();
+
+        let mut recv_snapshot = RocksSnapshot::new(&recv_snapshot_dir).unwrap();
+        recv_snapshot.write_all(&buf).await.unwrap();
+        // recv_snapshot.rewind().await.unwrap();
+        let engine_2 = RocksEngine::new(&recover_data_dir, &TESTTABLES).unwrap();
```
Comment (Phoenix500526) [inline engine/src/rocksdb_engine.rs:752]: Same as the `snapshot_should_work` test in `memory_engine.rs`

File: engine/src/engine_api.rs [hunk 9]
```diff
@@ -75,8 +79,17 @@ impl WriteOperation {
     }
 }
 
+/// This trait is a abstraction of the snapshot, We can Read/Write the snapshot like a file.
+pub trait SnapshotApi: AsyncSeek + AsyncWrite + AsyncSeek + Send + Sync {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/engine_api.rs]: my mistake, one of `AsyncSeek` should be `AsyncRead`, and in the new commit, we didn't use those traits

File: engine/src/rocksdb_engine.rs [hunk 10]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+    filename: String,
+    /// size of the file has been written
+    written_size: u64,
+    /// size of the file
+    size: u64,
+}
+
+impl SnapFile {
+    /// left size of this file
+    fn left_size(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: renamed

File: engine/src/rocksdb_engine.rs [hunk 11]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+#[derive(Debug, Default)]
+pub struct RocksSnapshot {
+    /// Snapshot meta
+    meta: Meta,
+    /// directory of the snapshot
+    dir: PathBuf,
+    /// files of the snapshot
+    snap_files: Vec<SnapFile>,
+    /// current file index
+    file_index: usize,
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

File: engine/src/rocksdb_engine.rs [hunk 12]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+            dir,
+            snap_files: Vec::new(),
+            file_index: 0,
+            current_file: None,
+            current_file_pos: 0,
+        })
+    }
+
+    /// Create a new snapshot for sending
+    fn new_for_sending<P>(dir: P) -> Result<RocksSnapshot, EngineError>
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

File: engine/src/rocksdb_engine.rs [hunk 13]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+    where
+        P: Into<PathBuf>,
+    {
+        let dir = dir.into();
+        let files = fs::read_dir(&dir)?;
+        let mut s = Self::new(dir)?;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

File: engine/src/rocksdb_engine.rs (hunk 6 above)
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

File: engine/src/rocksdb_engine.rs [hunk 14]
```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
...
+        }
+        let this = self.get_mut();
+        let (mut next_buf, mut written_bytes) = (buf, 0);
+        if this.meta.is_current {
+            let len = this.write_meta(next_buf)?;
+            written_bytes = written_bytes.overflow_add(len);
+            next_buf = &next_buf[len..];
+        }
+
+        while this.file_index < this.snap_files.len() {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: we need to write data to snapshot when follower receiving snapshot, and we can't write to a stream like this

File: engine/src/rocksdb_engine.rs [hunk 15]
```diff
@@ -60,7 +74,345 @@ impl RocksEngine {
...
+            snap_file_idx: 0,
+            current_file: None,
+        }
+    }
+
+    /// Create a new snapshot for recving
+    /// # Errors
+    /// Return `EngineError` when create directory failed.
+    #[inline]
+    pub fn new_for_recving<P>(dir: P) -> Result<Self, EngineError>
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs:160]: when follower need to receive a snapshot from leader, follower need to create a empty snapshot to write

File: engine/src/rocksdb_engine.rs [hunk 16]
```diff
@@ -60,7 +74,345 @@ impl RocksEngine {
...
+        RocksSnapshot {
+            meta: Meta::new(),
+            dir: dir.into(),
+            snap_files: Vec::new(),
+            snap_file_idx: 0,
+            current_file: None,
+        }
+    }
+
+    /// Create a new snapshot for recving
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs:156]: fixed

File: engine/src/rocksdb_engine.rs [hunk 17]
```diff
@@ -60,7 +74,345 @@ impl RocksEngine {
...
+    }
+}
+
+/// Meta data of the snapshot
+#[derive(Debug, Default)]
+struct Meta {
+    /// Meta data
+    data: Cursor<Vec<u8>>,
+    /// Whether the meta is current or not
+    is_current: bool,
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs:108]: changed

File: engine/src/rocksdb_engine.rs [hunk 18]
```diff
@@ -60,7 +74,345 @@ impl RocksEngine {
     }
 }
 
+/// meta of the snapshot
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: fixed

**Review by Phoenix500526 — APPROVED** [review]

**Review by markcty — APPROVED** [review]

**Pull request merged**